lib/pbench/server/api/resources/query_apis/iteration_samples.py

```diff
@@ -0,0 +1,240 @@
+        fields in the 'benchmark', 'sample', 'iterations' and other
+        sub-documents of result-data-sample index document that belong to the
+        given run id.
+        {
+            "run_id": "run id string"
+        }
+        json_data: JSON dictionary of type-normalized parameters
+            "run_id": String representation of run id
+        """
+        run_id = context.get("run_id")
+        dataset = context.get("dataset")
+
+        self.logger.info(
+            "Return sample iteration rows for user {}, prefix {}: for run id: {}",
+            Auth.token_auth.current_user().username,
+            self.prefix,
+            run_id,
+        )
+
+        # Retrieve the ES index that belongs to this run_id from the metadata
+        # table
+        try:
+            index_map = Metadata.getvalue(dataset=dataset, key="server.index-map")
+            for key, values in index_map.items():
+                if run_id in values:
+                    index = key
+                    self.logger.debug(f"Iteration samples index, {index!r}")
+        except MetadataError as e:
+            abort(HTTPStatus.BAD_REQUEST, message=str(e))
+
+        if index is None:
+            abort(
+                HTTPStatus.BAD_REQUEST,
+                message=f"Dataset with run_id: {run_id!r} not found",
+            )
+
+        try:
+            template = Template.find("result-data-sample")
+            mappings = template.mappings
+
+            # Only keep the whitelisted fields for aggregation
+            for key in mappings["properties"].copy():
+                if key not in IterationSamples.WHITELIST_AGGS_FIELDS:
+                    del mappings["properties"][key]
+        except TemplateNotFound:
+            self.logger.exception(
+                "Document template 'result-data-sample' not found in the database."
+            )
+            abort(HTTPStatus.NOT_FOUND, message="Mapping not found")
+
+        # Get all the fields from result-data-sample index that are of type keyword
+        result = self.get_keyword_fields(mappings)
+
+        # Build ES aggregation query for number of rows aggregations
+        aggs = {}
+        for key in result:
+            aggs[key] = {"terms": {"field": key}}
+
+        return {
+            "path": f"/{index}/_search",
+            "kwargs": {
+                "json": {
+                    "size": 0,
+                    "query": {"bool": {"filter": {"match": {"run.id": f"{run_id}"}}}},
+                    "aggs": aggs,
+                },
+                "params": {"ignore_unavailable": "true"},
+            },
+        }
+
+    def postprocess(self, es_json: JSON, context: CONTEXT) -> JSON:
+        """
+        Returns a non-empty list of aggregated unique values for keyword fields
+        in the 'benchmark', 'sample', 'iterations' and 'run' subdocuments of
+        result-data-sample index. The output only contains fields that has at
```

Comment:
Given that this API takes a run ID, it seems reasonable to me that it return results which are custom to that run.  I don't see how returning headers which have no values is going to be useful.  If the client really wants _that_, perhaps they should be querying the index mapping instead of a particular dataset.